### ISSUE TITLE
Lint for using `LazyLock` in a non-static context

### ISSUE THREAD

**Issue by tidely** [rust-lang/rust-clippy#15295]
### What it does

The lint should prevent users from accidentally assigning `LazyLock` using `let` instead of `static`. This should only apply when the `LazyLock` is initialized right away. This is pretty common within constructors for structs which need access to a common resource.

I'm aware of clippy having a feature freeze, but I'll likely just forget this unless I open it now

### Advantage

- Prevents accidentally using `LazyLock` in a way which re-initializes it each time

### Drawbacks

To prevent false positives we should only lint for `LazyLock`'s which are initialized right away, since they have uses even without `static` when used within structs

The users intent might also not be entirely clear, they might be using the `LazyLock` without `static` on accident, or they might not need it at all, but thought they did, however I think the latter case is much less common

### Example

```rust
fn get_platform() -> &'static str {
    let platform: LazyLock<&'static str> =
        LazyLock::new(|| get_platform_using_expensive_function());
    platform.as_ref()
}
```

Could be written as:

```rust
fn get_platform() -> &'static str {
    static PLATFORM: LazyLock<&'static str> =
        LazyLock::new(|| get_platform_using_expensive_function());
    PLATFORM.as_ref()
}
```

**Comment by kleinesfilmroellchen** [comment]
The exact same thing also happens with `const`, which originally was my intuition for how to use `LazyLock`. This lint could even be expanded to `const` + `LazyCell`, which equally performs a copy every time.

FWIW, I just had a program spend 1/3 of its runtime in `const LazyCell` initialization. So this is very much a real issue.
